src/ILCompiler.Compiler/src/Compiler/TypeExtensions.cs

```diff
@@ -79,7 +79,7 @@ public static bool RequiresInstMethodTableArg(this MethodDesc method)
         /// </summary>
         public static bool AcquiresInstMethodTableFromThis(this MethodDesc method)
         {
-            return method.IsSharedByGenericInstantiations &&
+            return method.IsCanonicalMethod(CanonicalFormKind.Any) &&
```

Comment:
I eagerly stole `IsSharedByGenericInstantiations` from NUTC, but it might be something specific to the UTC inliner. It means: is the method (or the owning type) instantiated over canonical types (either a __Canon, or a struct instantiated over __Canon).

In particular `IsSharedByGenericInstantiations` will say false on, say, `Foo<System.__Canon[]>` which does show up sometimes on the JitInterface side before it gets normalized. Alternative approach would be to be very careful about when things are normalized or aren't.